### ISSUE TITLE
Add ability to add Annotations in web uploader

### ISSUE THREAD

**Issue by tiwane** [inaturalist/inaturalist#2353]
From: https://forum.inaturalist.org/t/add-annotations-when-creating-observations/1513/10

Users are asking to have an Annotations section on the uploader page. This is especially useful for those who have slow internet connections, and they can annotate the observations while the photos and metadata are loading.

It could be an option on the left-hand side, by the Projects, Tags, Obs Fields, etc. Only Attribute and Value columns are needed, and only Attributes and Values which correspond to the species guess of the observation.

**Comment by tiwane** [comment]
@kueda and @pleary would this be difficult to implement?

**Comment by kueda** [comment]
Not impossible but also not totally trivial.

**Comment by todtb** [comment]
Hey @tiwane, I would love to take this on as I also would find it quite useful. Even more so in the Android App, but that's a bit out of my expertise at the moment. Is this still on the table?

Also @kueda, when you said "not totally trivial", I suspect part of the issue is pulling proper annotations on the fly for the species in question. Are there other foreseeable challenges here to be aware of, such as something with the data model? I see that an annotation `belongs_to` a resource but is validating that it can only be of the type `Observation`. Maybe something else?

Basically, I don't want to get too deep into investigating if you think there may be some major prerequisites here.

**Comment by tiwane** [comment]
@todtb I'd love it but leave it to @kueda for the go-ahead. FWIW I'm not a dev, I basically relay bug reports and requests from the community.

Thanks for all of your work, btw!

**Comment by kueda** [comment]
The not trivial part would be either attempting to integrate our [existing annotations component](https://github.com/inaturalist/inaturalist/blob/master/app/webpack/observations/show/components/annotations.jsx) (and maybe the [duck](https://github.com/inaturalist/inaturalist/blob/master/app/webpack/observations/show/ducks/controlled_terms.js)) from obs detail into the uploader, building a new one, or choosing the right balance between the two, and dealing with our controlled terms system which can be a bit complicated. Fetching the right controlled terms is actually pretty easy once you know what taxon you're fetching them for ([example](https://api.inaturalist.org/v1/controlled_terms/for_taxon?taxon_id=48460%2C1%2C47120%2C372739%2C47158%2C184884%2C47201%2C124417%2C326777%2C48740%2C52747%2C119344%2C152993)), but yes, getting the right controlled terms for a variety of observations associated with a variety of taxa presents some UI challenges, if not implementation challenges. And if you're not going to reuse or adapt that component, there's some annoying logic in controlled terms like "blocking" values whose presence makes it impossible to add other values (e.g. "No Evidence of Flowering" in "Plant Phenology").

If I was going to work on this, I guess I'd want answers to

1. What does the side bar annotation creation UI look like?
1. What does the side bar annotation editing UI look like, or do you just delete annotations you've added so you can add another?
1. What does the sidebar look like when you have multiple cards selected and they have multiple annotations?
1. What annotations can I add when I select multiple cards associated with multiple taxa? Just the annotations relevant to the taxon that contains them all? How can the UI indicate that I would have more options if I just selected taxa representing a smaller subtree (e.g. just plants)?

Personally, I'd be happy with muddling through these and figuring out what works as I build it out (much to @abhasinat's horror). If you want to do the same, go for it, but understand that I can't guarantee we'll like or integrate your solution. I guess that's true of any OS development, but I find UI work to be particularly time consuming.

**Comment by todtb** [comment]
Thanks for the quick response. Excellent points and ones I too do not have the answers to. Honestly, I hadn't really thought long enough about just how different the mobile and web interfaces are in functionality. 

Particularly challenging I think are:

> 3) What does the sidebar look like when you have multiple cards selected and they have multiple annotations?
> 4) What annotations can I add when I select multiple cards associated with multiple taxa? Just the annotations relevant to the taxon that contains them all? How can the UI indicate that I would have more options if I just selected taxa representing a smaller subtree (e.g. just plants)?

I suppose it's possible to exclude taxon specific fields such as annotations when multiple observations are selected, but then that introduces an inconsistency in the UI and one that may not be immediately intuitive. 

---
> The not trivial part would be either attempting to integrate our existing annotations component (and maybe the duck) from obs detail into the uploader, building a new one

> 1) What does the side bar annotation creation UI look like?

I was picturing a new component that would match the existing uploader sidebar's aesthetic for the sake of simplicity and consistency, however would never assume to know what would be the best/final approach to take on UI.


---
> If you want to do the same, go for it, but understand that I can't guarantee we'll like or integrate your solution.

Frankly, I don't know the community/vision for the app well enough at this point to make an educated guess on a good solution that fits all needs. Therefore I think until we have the those answers and can come up with more granular tasks, I should probably hold off.

**Comment by tiwane** [comment]
I'll see if I can get some feedback from the community.

**Comment by kueda** [comment]
Please don't ask for more community feedback. This needs design and decision, not more ideas.

**Comment by tiwane** [comment]
>What does the side bar annotation creation UI look like?

I think matching the sidebar's aesthetic would be best, but interested if @abhasinat has any thoughts. We'd need to come up with an icon like the other elements have...maybe a checkmark? I don't think we have an existing one for annotations.

> What does the side bar annotation editing UI look like, or do you just delete annotations you've added so you can add another?

I'm not sure I understand this, can you provide a specific example?

> What does the sidebar look like when you have multiple cards selected and they have multiple annotations?

I understand the consistency issue here, but I think it would be fine to not display the annotation picker if multiple cards are selected. The functionality was requested mainly be people who have slow connections and time on their hands while waiting for everything to upload.

**Comment by kueda** [comment]
> > What does the side bar annotation editing UI look like, or do you just delete annotations you've added so you can add another?
> I'm not sure I understand this, can you provide a specific example?

The Uploader supports adding new info and editing existing info. For observation field values, for example, you cannot edit an OFV that you added, only delete it. Annotations are even more complicated, though, because they support voting, and the potential for multiple annotations for the same attribute term. On obs detail, you can't edit the annotation, really, but you can vote on it, so are we going to support voting in the Uploader? IMO, no.

> > What does the sidebar look like when you have multiple cards selected and they have multiple annotations?
> I understand the consistency issue here, but I think it would be fine to not display the annotation picker if multiple cards are selected.

I'll take your word for it, but if I were someone with a slow connection uploading a lot of plants that I wanted to annotate as flowering, I'd be pretty annoyed that I couldn't select them all and add that annotation in one go.

**Comment by tiwane** [comment]
> so are we going to support voting in the Uploader? IMO, no.

Agreed, add or remove the annotation.

> I'll take your word for it, but if I were someone with a slow connection uploading a lot of plants that I wanted to annotate as flowering, I'd be pretty annoyed that I couldn't select them all and add that annotation in one go.

I hear you. Seems like the difficulty here is making it clear which annotations are available (or not available) with the current selection? Figuring out which annotations (if any) _do_ fit all selected observations is not difficult?

**Comment by tiwane** [comment]
Bringing this back up as people are still voting for it and I think the ball's in my court and it's been there for way too long. Ideally, if someone selects a batch of flowering plants, it would be cool if they could annotate them all at once. 

If someone selects a batch that has observations of both plants and animals, are we able to show the annotations which fit them all (eg `Sex`)? If a batch contains unidentified observations, should we just gray out the Annotation option?

**Comment by kueda** [comment]
> If someone selects a batch that has observations of both plants and animals, are we able to show the annotations which fit them all (eg Sex)?

I think that should be possible.

> If a batch contains unidentified observations, should we just gray out the Annotation option?

I think that makes sense, but I defer to @abhasinat 

Another annoying wrinkle is what to do when you have an obs of a plant, add a Plant Phenology annotation, then change the taxon for that obs to an insect. Is the front end supposed to decide that the annotation is now invalid, or do we deal with that when we try to post the annotation, which will be invalid?

**Comment by abhasinat** [comment]
I'm fine with graying out the annotation option if there are unidentified observations, but I would also include some error text there if a user tries to click the grayed-out annotations button, like "Annotations can only be added to observations with an identification"
